### ISSUE TITLE
Fix `media-feature-name-no-unknown` false positives for `environment-blending`

### ISSUE THREAD

**Issue by Mouvedia** [stylelint/stylelint#6968]
### What steps are needed to reproduce the bug?

```css
@media (min--moz-device-pixel-ratio: 1.3) {}
@media (screen-spanning: single-fold-vertical) {}
@media (environment-blending: additive) {}
/* etc. */
```

### What Stylelint configuration is needed to reproduce the bug?

```json
{
  "rules": {
    "media-feature-name-no-unknown": true
  }
}
```

### How did you run Stylelint?

https://stylelint.io/demo/#N4Igxg9gJgpiBcIACBbGUCWBDABAChQwDsBaElCALxNgDcMwYSAHDADxgBsSAnLAFwwR4OAIwA6AMwBKHMAC+AHSKp02fAGcwPGDFIbmWIkWIBzERrOcmAMwicoJWjB6CwWTrIXLVmXHj16HggiNCJ+EgAjayJMInMcLChMQWcveRAAGhAbDGsAOSw0BBAYNiLma3EwDQ0s8BDc0xLgZRwcRRAeAFdrDU6RVqJ29s60PxIbGAFunRIiIqYiCBJuogBrZYB3IgGcfh6YNpwlIgzsyCImgDEIHhQBEoArDRD62GY6xCHRkA1+ACe1k4xH4e06nAEMH+nUyx06-yBXFBJEuTRI-yMUCwPCg4JAkP40LBIGUGXkQA

### What did you expect to happen?

No errors.

### What actually happened?

Unexpected unknown media feature name "placeholder"

### Resources

- https://www.w3.org/TR/mediaqueries-5/#environment-blending
- https://browserleaks.com/css
- https://github.com/postcss/autoprefixer/issues/122#issuecomment-49451372
- edge requires `enable-experimental-web-platform-features` flags

**Comment by ybiquitous** [comment]
@Mouvedia Thanks for the report. Can you please provide a *reproducible* demo, not just `stylelint.io/demo`? Also, please provide links to the CSS spec resource?

**Comment by Mouvedia** [comment]
@ybiquitous Iv updated _How did you run Stylelint?_ and added _ Resources_.

**Comment by romainmenke** [comment]
I think these are three different cases :
- `min--moz-device-pixel-ratio` is a vendor prefixed non-standard media feature
- `screen-spanning` is a non-standard media feature
- `environment-blending` is a standard media feature

----------

`min--moz-device-pixel-ratio` is vendor prefixed in an unexpected way.

see : https://developer.mozilla.org/en-US/docs/Web/CSS/@media/-moz-device-pixel-ratio

> Note: This media feature is also implemented by WebKit as -webkit-device-pixel-ratio. The min and max prefixes as implemented by Gecko are named min--moz-device-pixel-ratio and max--moz-device-pixel-ratio; but the same prefixes as implemented by Webkit are named -webkit-min-device-pixel-ratio and -webkit-max-device-pixel-ratio.

We should handle `min--moz` the same as `-webkit-min`.  
Maybe a new utility to easily implement this in all relevant rules?

-----

`screen-spanning` isn't part of any specification or draft specification I could find.

I don't think it should be added to the known media features.  
I think it's best to add it with the `ignoreMediaFeatureNames` option.

-----

`environment-blending` seems to be missing here : https://github.com/stylelint/stylelint/blob/main/lib/reference/mediaFeatures.js#L29

Are there others that are missing?

**Comment by Mouvedia** [comment]
> Maybe a new utility to easily implement this in all relevant rules?

Let's start small: this rule and 2 exceptions for now.

> I don't think it should be added to the known media features.

I disagree. IMO if a browser at one point in time supported a media-feature-name, it should be considered as known.

> Are there others that are missing?

Yes there are, and if someone takes the responsibility to fix this issue he will have to be thorough.
I might do it eventually do it if no one does in the mean time.

**Comment by romainmenke** [comment]
> I disagree. IMO if a browser at one point in time supported a media-feature-name, it should be considered as known.

That would deviate from how the rule is documented today : 

https://stylelint.io/user-guide/rules/media-feature-name-no-unknown

> This rule considers media feature names defined in the CSS Specifications, up to and including Editor's Drafts, to be known.

Maybe it can be a rule option?

**Comment by ybiquitous** [comment]
I agree with fixing false positives for `environment-blending` because it's a spec.

On the other hand, I think users can ignore the other two names by using the `ignoreMediaFeatureNames` option because they're not standard.

If users complain about non-standard names in the future, then we can revisit them.

**Comment by Mouvedia** [comment]
`min--moz-device-pixel-ratio` and `max--moz-device-pixel-ratio` are comparable to `-webkit-min-device-pixel-ratio` and `-webkit-max-device-pixel-ratio` which are currently ignored, they should too.

> That would deviate from how the rule is documented today

I am a pragmatic and care about cross-browser compatibility. If I did a PR Id add all the missing ones because from the user perspective it would be ideal, we already have `media-feature-name-no-vendor-prefix` for the ones that care about that kind of things.

---

off topic

> We should handle `min--moz` the same as `-webkit-min`.

@romainmenke `media-feature-name-no-vendor-prefix` has a minor bug: the error start range is off by 3.
i.e. reports min<i>--moz-device-pixel-ratio</i> instead of *min--moz-device-pixel-ratio*

**Comment by romainmenke** [comment]
@romainmenke said :

> That would deviate from how the rule is documented today

@Mouvedia said :

> I am a pragmatic and care about cross-browser compatibility. If I did a PR Id add all the missing ones because from the user perspective it would be ideal, we already have media-feature-name-no-vendor-prefix for the ones that care about that kind of things.

My comment was specifically about `screen-spanning`.

The rule itself is documented as only considering those media features valid that are part of a specification. This feature is not part of any specification or draft.

Users who read the documentation will expect a warning.  
I expect a warning for `screen-spanning` :)

There is also an escape hatch for those who are using `screen-spanning`.  
They can set `ignoreMediaFeatureNames`.

If however `screen-spanning` is added to the list of known media features, there is no way to bring back the warning for those who expect the documented behavior.

-------

> `media-feature-name-no-vendor-prefix` has a minor bug: the error start range is off by 3.
i.e. reports `min--moz-device-pixel-ratio` instead of `min--moz-device-pixel-ratio`

I am unsure what you mean. The two strings you mention are exact equals. :)  
Maybe best to open a separate issue for this?

**Comment by Mouvedia** [comment]
> The two strings you mention are exact equals. :)

Check the italic style of my comment or try it in stylelint.io/demo and check the text-decoration-style of the first three letters.

**Comment by romainmenke** [comment]
I've looked a bit more into `min--moz-device-pixel-ratio` and `max--moz-device-pixel-ratio`.

This in itself is quite interesting because it was only very briefly useful in a browser that never had much adoption. And yet people keep using it.

Web Almanac : https://almanac.httparchive.org/en/2022/css#responsive-design

Used in ±10%

<img width="677" alt="Screenshot 2023-06-25 at 13 12 01" src="https://github.com/stylelint/stylelint/assets/11521496/b6ae531d-2705-4dd2-85cb-bd66b5b07500">

Caniuse : https://caniuse.com/?search=resolution

Autoprefixer only adds it when targeting Firefox 15 or older : 
https://preset-env.cssdb.org/playground/#eyJzb3VyY2UiOiJAbWVkaWEgKG1pbi1yZXNvbHV0aW9uOiAxZHBpKSB7XG4gIGRpdiB7XG4gICAgY29sb3I6IGdyZWVuO1xuICB9XG59IiwiY29uZmlnIjp7ImJyb3dzZXJzIjpbImZpcmVmb3ggPj0gMTUiXSwibWluaW11bVZlbmRvckltcGxlbWVudGF0aW9ucyI6MCwic3RhZ2UiOjIsImxvZ2ljYWwiOnsiaW5saW5lRGlyZWN0aW9uIjoibGVmdC10by1yaWdodCIsImJsb2NrRGlyZWN0aW9uIjoidG9wLXRvLWJvdHRvbSJ9fX0%3D

Are people adding this manually, if so, why?

----

From a technical point of view it is trivial to change the vendor prefix utility to have support for this one exception.

```js
const VENDOR_PREFIX_MATCH = /^(?:min-|max-)?(-\w+-)/i;
const VENDOR_PREFIX_REPLACE = /^(min-|max-)?(-\w+-)/i;

const VENDOR_PREFIX_REPLACER = (_match, p1) => {
	return p1 ?? '';
}

module.exports = {
	prefix(prop) {
		const match = prop.match(VENDOR_PREFIX_MATCH);

		if (match) {
			return match[1] || '';
		}

		return '';
	},

	unprefixed(prop) {
		return prop.replace(VENDOR_PREFIX_REPLACE, VENDOR_PREFIX_REPLACER);
	},
};
```

But I don't think it is worth it.

This was clearly a one time mistake in Gecko.
This mistake was rectified [11 years ago](https://github.com/mdn/browser-compat-data/blob/main/browsers/firefox.json#L131).

I agree with @ybiquitous that we can always revisit later if people complain about this again.

**Comment by romainmenke** [comment]
> `media-feature-name-no-vendor-prefix` has a minor bug: the error start range is off by 3.
i.e. reports min--moz-device-pixel-ratio instead of min--moz-device-pixel-ratio

see : https://github.com/stylelint/stylelint/issues/6975

**Comment by jeddy3** [comment]
@romainmenke Thank you for the informative deep dive.

> I agree with @ybiquitous that we can always revisit later if people complain about this again.

SGTM. Let's limit this issue to `environment-blending`.
